src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs

```diff
@@ -47,43 +70,176 @@ protected override void EndProcess ()
+				return;
+			}
+
+			// Clear IL in method body
+			method.Body.Instructions.Clear ();
 
-		Context.LogMessage ($"Writing {TypeMappings.Count} typemap entries");
-		var il = method.Body.GetILProcessor ();
-		var addMethod = module.ImportReference (typeof (IDictionary<string, Type>).GetMethod ("Add"));
-		var getTypeFromHandle = module.ImportReference (typeof (Type).GetMethod ("GetTypeFromHandle"));
-		foreach (var (javaName, list) in TypeMappings) {
-			/*
-			 * IL_0000: ldarg.0
-			 * IL_0001: ldfld class [System.Runtime]System.Collections.Generic.IDictionary`2<string, class [System.Runtime]System.Type> Microsoft.Android.Runtime.NativeAotTypeManager::TypeMappings
-			 * IL_0006: ldstr "java/lang/Object"
-			 * IL_000b: ldtoken [Mono.Android]Java.Lang.Object
-			 * IL_0010: call class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
-			 * IL_0015: callvirt instance void class [System.Runtime]System.Collections.Generic.IDictionary`2<string, class [System.Runtime]System.Type>::Add(!0, !1)
-			 */
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldarg_0);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldfld, field);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldstr, javaName);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Ldtoken, module.ImportReference (SelectTypeDefinition (javaName, list)));
-			il.Emit (Mono.Cecil.Cil.OpCodes.Call, getTypeFromHandle);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Callvirt, addMethod);
+			var il = method.Body.GetILProcessor ();
+
+			var orderedMapping = TypeMappings.Select(kvp => (Hash (kvp.Key), SelectTypeDefinition (kvp.Key, kvp.Value))).OrderBy (x => x.Item1);
+
+			var targets = new List<Instruction> ();
+			foreach (var name in names) {
+				targets.Add (il.Create (OpCodes.Ldstr, name));
+			}
+
+			il.Emit (OpCodes.Ldarg_0);
+			il.Emit (OpCodes.Switch, targets.ToArray ());
+
+			var defaultTarget = il.Create (OpCodes.Ldnull);
+			il.Emit (OpCodes.Br, defaultTarget);
+
+			foreach (var target in targets) {
+				il.Append (target);
+				il.Emit (OpCodes.Ret);
+			}
+
+			il.Append (defaultTarget);
+			il.Emit (OpCodes.Ret);
 		}
 
-		il.Emit (Mono.Cecil.Cil.OpCodes.Ret);
+		void GenerateHashesFieldInitialization (ulong[] hashes)
+		{
+			// Sanity check: hashes must be unique and sorted
+			if (hashes.Length > 0) {
+				ulong previous = hashes[0];
+				for (int i = 1; i < hashes.Length; ++i) {
+					if (hashes[i] == previous) {
+						throw new InvalidOperationException ($"Duplicate hashes");
+					} else if (hashes[i] < previous) {
+						throw new InvalidOperationException ($"Hashes are not in ascending order");
+					}
+
+					previous = hashes[i];
+				}
+			}
+
+			var privateImplementationDetails = module.Types.FirstOrDefault (t => t.Name.Contains ("<PrivateImplementationDetails>"));
+			if (privateImplementationDetails is null) {
+				Context.LogMessage ($"Unable to find <PrivateImplementationDetails> class");
+				return;
```

Comment:
There are a couple places like this that log and `return` early. Should they `throw` instead?

The issue is you'd need `-p:_ExtraTrimmerArgs=--verbose` to see the message in a `.binlog`.

Reply:
This should probably throw. On the other hand, we should be able to introduce the class if it isn't there yet. I will update the code.